hazelcast/src/main/java/com/hazelcast/spi/impl/UnmodifiableLazyList.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.SerializationService;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+/**
+ * TODO add a proper JavaDoc
+ */
+public class UnmodifiableLazyList<E> implements List<E> {
```

Comment:
actually it is the opposite, when deserilaized first, I replace the serialized version. This way you don't need to deserialize again and you don't get different instances.

PS: I will add more tests and javadoc